test/test_connectionpool.py

```diff
@@ -389,6 +389,96 @@ def _make_request(self, *args, **kwargs):
+                    if cls.reject_buffering_arg:
+                        cls.rejected_call_count += 1
+                        raise TypeError("Buffering argument not supported!")
+                response = httplib.HTTPResponse(MockSock)
+                response.fp = MockChunkedEncodingResponse([b'f', b'o', b'o'])
+                response.msg = HTTPMessage()
+                response.headers = HTTPHeaderDict()
+                return response
+
+        class CustomResponse(HTTPResponse):
+            @classmethod
+            def from_httplib(cls, httplib_response, pool, connection, **kwds):
+                return cls(pool=pool, connection=connection)
+
+        class CustomConnectionPool(HTTPConnectionPool):
+            ConnectionCls = CustomConnection
+            ResponseCls = CustomResponse
+
+            def _validate_conn(self, conn):
+                return True
+
+        # Set up a custom pool that lets us test the getresponse() handling
+        pool = CustomConnectionPool(host='localhost', maxsize=1, block=True)
+        def _make_new_request(**request_kwds):
+            return pool.request('GET', '/',
+                                redirect=False,
+                                retries=False,
+                                **request_kwds,
+            )
+        # First check the behaviour with the buffering argument accepted
+        # This also checks that the custom connection and response are used
+        response = _make_new_request(release_conn=False)
+        self.assertIsInstance(response, CustomResponse)
+        self.assertIsInstance(response.connection, CustomConnection)
+        self.assertEqual(CustomConnection.explicitly_buffered_call_count, 1)
+        self.assertEqual(CustomConnection.rejected_call_count, 0)
+        response.release_conn()
+        # Then check the first time it fails
+        CustomConnection.reject_buffering_arg = True
+        response = _make_new_request()
+        self.assertEqual(CustomConnection.explicitly_buffered_call_count, 2)
+        self.assertEqual(CustomConnection.rejected_call_count, 1)
+        # Then check the argument isn't even tried on subsequent calls
+        response = _make_new_request()
+        self.assertEqual(CustomConnection.explicitly_buffered_call_count, 2)
+        self.assertEqual(CustomConnection.rejected_call_count, 1)
+        # Finally, check a completely broken custom pool raises TypeError
+        class BadConnectionPool(HTTPConnectionPool):
+            ConnectionCls = CustomConnection
+            ResponseCls = CustomResponse
+
+            def _validate_conn(self, conn):
+                self.ConnectionCls = HTTPConnection
+                return True
+        bad_pool = BadConnectionPool(host='localhost', maxsize=1, block=True)
+        error_structure = 'expected.*HTTPConnection.*got.*CustomConnection'
+        with self.assertRaisesRegex(TypeError, error_structure):
+            bad_pool.request('GET', '/', redirect=False, retries=False)
+
```

Comment:
This test is extremely complex, mostly because it defines a few helper objects inline which makes them unavailable for reuse in granular tests. It might be better to split this test out to a new file, define the helpers at module scope, and then write a collection of smaller tests.